ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java

```diff
@@ -28,46 +41,90 @@
+   * Construct an Azure cloud destination from config properties.
+   * @param verProps the {@link VerifiableProperties} to use.
    * @throws InvalidKeyException if credentials in the connection string contain an invalid key.
    * @throws URISyntaxException if the connection string specifies an invalid URI.
    */
-  AzureCloudDestination(String configSpec) throws URISyntaxException, InvalidKeyException {
-    this(CloudStorageAccount.parse(configSpec));
+  AzureCloudDestination(VerifiableProperties verProps) throws URISyntaxException, InvalidKeyException {
+    String configSpec = verProps.getString(STORAGE_CONFIG_SPEC);
+    azureAccount = CloudStorageAccount.parse(configSpec);
+    azureBlobClient = azureAccount.createCloudBlobClient();
+    String cosmosServiceEndpoint = verProps.getString(COSMOS_ENDPOINT);
+    String cosmosKey = verProps.getString(COSMOS_KEY);
+    cosmosCollectionLink = verProps.getString(COSMOS_COLLECTION_LINK);
+    documentClient =
+        new DocumentClient(cosmosServiceEndpoint, cosmosKey, ConnectionPolicy.GetDefault(), ConsistencyLevel.Session);
+    // check that it works
+    try {
+      ResourceResponse<DocumentCollection> response =
+          documentClient.readCollection(cosmosCollectionLink, defaultRequestOptions);
+      if (response.getResource() == null) {
+        throw new IllegalArgumentException("CosmosDB collection not found: " + cosmosCollectionLink);
+      }
+    } catch (DocumentClientException ex) {
+      throw new IllegalArgumentException("Invalid CosmosDB properties", ex);
+    }
+    logger.info("Created Azure destination");
   }
 
   /**
-   * Construct an Azure cloud destination from a {@link CloudStorageAccount} instance.
+   * Test constructor.
    * @param azureAccount the {@link CloudStorageAccount} to use.
+   * @param documentClient the {@link DocumentClient} to use.
+   * @param cosmosCollectionLink the CosmosDB collection link to use.
    * @throws CloudStorageException if the destination could not be created.
    */
-  AzureCloudDestination(CloudStorageAccount azureAccount) {
+  AzureCloudDestination(CloudStorageAccount azureAccount, DocumentClient documentClient, String cosmosCollectionLink) {
     this.azureAccount = azureAccount;
+    this.documentClient = documentClient;
+    this.cosmosCollectionLink = cosmosCollectionLink;
 
     // Create a blob client to interact with Blob storage
     azureBlobClient = azureAccount.createCloudBlobClient();
-    LOGGER.info("Created Azure destination");
   }
 
-  // TODO: add blob properties arg in order to set Azure blob metadata
+  /**
+   * For integration test
+   * @return the CosmosDB DocumentClient.
+   */
+  DocumentClient getDocumentClient() {
+    return documentClient;
+  }
+
   @Override
-  public boolean uploadBlob(BlobId blobId, long blobSize, InputStream blobInputStream) throws CloudStorageException {
+  public boolean uploadBlob(BlobId blobId, long blobSize, CloudBlobMetadata cloudBlobMetadata,
```

Comment:
Rather than making `CloudDestination` aware of `BlobId`, do you think it is better for the caller to parse the id string and container and pass it in?

Reply:
I originally had it as string, but realized that every method requires extracting the blob's partition id (to know where to find it in both storage and metadata).  While uploadBlob() takes CloudBlobMetadata which contains everything, I didn't want to burden the udpate and delete methods with this extra parameter just to specify the partitionId.  I figured it was okay to pass BlobId since the destination logic has to have some knowledge of Ambry-isms anyway.